### DISCUSSION TITLE
Filtering on empty relation

### DISCUSSION THREAD

**Discussion by onordgren** [directus/directus#11608] [Archive]
Hey! I have a basic many to many relation, for example a recipes collection and an ingredients collection with a joining table.

This is working fine but I would like to be able to filter out all of the recipes that doesn't have any ingredients.

I have tried _null, _empty on both the ingredients alias as well as the nested id but it doesn't really work. The opposite, i.e. only fetching the ones with ingredients, works fine with _nnull for example.

Any ideas?

**Comment by xu4wang** [comment]
I did some test on my local machine, here are some findings.

The configuration:  there are 3 collections, Products, Tags and Tenants. 
- Products and Tags are with M2M relationship.
- Products has a M2O field tenant_id pointing to Tenants.

For M2M field, the search/filter is on the junction collection actually. Below queries return the same result, say all the products with tags.

```
http://localhost:8055/items/Products?filter[tags][Products_id][_nnull]=true
http://localhost:8055/items/Products?filter[tags][Tags_id][_nnull]=true
http://localhost:8055/items/Products?filter[tags][id][_nnull]=true


{
 "data": [
   {
     "id":1,
     ...
   }
   ...
 ]
}
```

And below queries all return empty result. I was also not able to get the products without tag.
I guess the reason is in the junction table, there is no items selected by the query.

```
http://localhost:8055/items/Products?filter[tags][Products_id][_null]=true
http://localhost:8055/items/Products?filter[tags][tags_id][_null]=true
http://localhost:8055/items/Products?filter[tags][id][_null]=true

{
 "data": []
}
```

For M2O field, I was able to use both _null and _nnull.  

```
http://localhost:8055/items/Products?filter[tenant_id][_null]=true      //return products does not belong to any company.
http://localhost:8055/items/Products?filter[tenant_id][_nnull]=true     //return products belongs to one company.
http://localhost:8055/items/Products?filter[tenant_id][name][_eq]=Company%20A //return the Products of "Company A".
```

**Reply by onordgren** [reply]
Cheers @xu4wang - these are exactly my findings as well. I suppose a workaround for a client is to first call the junction table and get a list of all the ids of the source collection. Then do another request to the source collection and filter away the ids that was in the junction.

I would however like to do this in a field so a user can only see/select recipes that don't have any ingredients in them. Seems like its more tricky than I thought :)

**Comment by xu4wang** [comment]
Yes.  Your workaround should work...

There are _and or _or logical operators for now. Maybe in the future a _not operator is a good one for this scenario.


This was also discussed here #10695

**Reply by kirkstrobeck** [reply]
This was my solve

```graphql
query {
    my_big_collection(filter: {
      m2o_thing_here: {
        id: {
          _null: true
        }
      }
    }) {
      name
    }
}
```

**Reply by luciandavidescu** [reply]
Trying to filter out `last_page` in_directus users, with flows, `_nnull` set to either true or false is the only thing that (kinda) works in making a distinction, but they both return _true_ when `last_page` is present, so I'd have to follow up on the failure path.

```
{
    "$trigger": {
        "payload": {
            "last_page": {
                "_nnull": false
            }
        }
    }
}
```

**Comment by azrikahar** [comment]
This is now made possible thanks to #12488 👍 So you can do something like `?filter[count(your_related_field)][_eq]=0` for ones _without_ related items.

**Reply by onordgren** [reply]
That's awesome! Do you know if it's possible to do in the app as well? For example to be able to filter out which items to show in a one to many relation drop down based on this value.

**Comment by benhaynes** [comment]
These functions (COUNT) are also in the app, available in filters and validations.

**Reply by onordgren** [reply]
Super!
